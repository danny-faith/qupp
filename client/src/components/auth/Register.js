import React, { Component } from 'react'
import { Row, Col, Button, Input } from 'react-materialize';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { registerUser } from '../../actions/authActions';
import classnames from 'classnames';

class Register extends Component {
	state = {
		username: '',
		email: '',
		password: '',
		password2: '',
		errors: {}
	}
	onChange = e => {
		this.setState({[e.target.name]: e.target.value});
	}
	onSubmit = e => {
		e.preventDefault();
		const newUser = {
			username: this.state.username,
			email: this.state.email,
			password: this.state.password,
			password2: this.state.password2
		}
		this.props.registerUser(newUser, this.props.history);
	}
	componentDidMount = () => {
        if (this.props.auth.isAuthenticated) {
            this.props.history.push('/dashboard');
        }
    }
	componentWillReceiveProps = (nextProps) => {
		if (nextProps.errors) {
			this.setState({errors: nextProps.errors});
		}
	}
	render() {
		const { errors } = this.state;
		return (
			<Row>
				<Col s={6} className="offset-s3">
					<h1 className="center">Register</h1>
					<form noValidate onSubmit={this.onSubmit}>
						<Row>
							<Col s={12}>
								<Input
									id={"username"}
									className={classnames({
										'invalid': errors.username
									})} 
									type="text"
									name="username"
									placeholder="Username"
									s={12}
									label="Username"
									onChange={this.onChange}
									value={this.state.username}
									/>
									{errors.username && (<p className="red-text col no-margin">{errors.username}</p>)}
							</Col>
						</Row>
						<Row>
							<Col s={12}>
								<Input
									id={"email"}
									className={classnames({
										'invalid': errors.email
									})} 
									type="email"
									name="email"
									placeholder="Email"
									s={12}
									label="Email"
									onChange={this.onChange}
									value={this.state.email}
									/>
									{errors.email && (<p className="red-text col no-margin">{errors.email}</p>)}
							</Col>
						</Row>
						<Row>
							<Col s={12}>
								<Input
									id={"password"}
									className={classnames({
										'invalid': errors.password
									})} 
									type="password"
									name="password"
									placeholder="Password"
									s={12}
									label="Password"
									onChange={this.onChange}
									value={this.state.password}
									/>
									{errors.password && (<p className="red-text col no-margin">{errors.password}</p>)}
							</Col>
						</Row>
						<Row>
							<Col s={12}>
								<Input
									id={"password2"}
									className={classnames({
										'invalid': errors.password2
									})} 
									type="password"
									name="password2"
									placeholder="Confirm password"
									s={12}
									label="Confirm password"
									onChange={this.onChange}
									value={this.state.password2}
									/>
									{errors.password2 && (<p className="red-text col no-margin">{errors.password2}</p>)}
							</Col>
						</Row>
						<Row>
							<Col s={12}>
								<Button className="btn-small right" waves="light">Register</Button>
							</Col>
						</Row>
					</form>
				</Col>
			</Row>
			)
		}
}

Register.propTypes = {
	registerUser: PropTypes.func.isRequired,
	auth: PropTypes.object.isRequired,
	errors: PropTypes.object.isRequired,
}

const mapStateToProps = (state) => ({
	auth: state.auth,
	errors: state.errors
});

export default connect(mapStateToProps, {registerUser })(withRouter(Register));