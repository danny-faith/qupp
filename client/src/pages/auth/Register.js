import React, { Component } from 'react'
import { Row, Col, Button } from 'react-materialize';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { registerUser } from '../../actions/authActions';
import TextFieldGroup from '../../components/common/TextFieldGroup';

export class Register extends Component {
	state = {
		username: '',
		email: '',
		password: '',
		password2: '',
		errors: {}
	}
	onChange = e => {
		const errors = this.state.errors;
		errors[e.target.name] = '';
		this.setState({
			[e.target.name]: e.target.value,
			errors
		});
	}
	onSubmit = e => {
		e.preventDefault();
		const newUser = {
			username: this.state.username,
			email: this.state.email,
			password: this.state.password,
			password2: this.state.password2
		}
		this.props.registerUser(newUser, this.props.history);
	}
	componentDidMount = () => {
        if (this.props.auth.isAuthenticated) {
            this.props.history.push('/dashboard');
        }
    }
	componentWillReceiveProps = (nextProps) => {
		if (nextProps.errors) {
			this.setState({errors: nextProps.errors});
		}
	}
	render() {
		const { errors } = this.state;
		return (
			<Row>
				<Col s={6} className="offset-s3">
					<h1 className="center">Register</h1>
					<form noValidate onSubmit={this.onSubmit}>
						<Row>
							<TextFieldGroup
								name="email"
								type="email"
								label="Email"
								value={this.state.email}
								onChange={this.onChange}
								error={errors.email}
              				/>
							<TextFieldGroup
								name="username"
								label="Username"
								value={this.state.username}
								onChange={this.onChange}
								error={errors.username}
              				/>
							<TextFieldGroup
								name="password"
								type="password"
								label="Password"
								value={this.state.password}
								onChange={this.onChange}
								error={errors.password}
              				/>
							<TextFieldGroup
								name="password2"
								type="password"
								label="Confirm password"
								value={this.state.password2}
								onChange={this.onChange}
								error={errors.password2}
              				/>
						</Row>
						<Row>
							<Col s={12}>
								<Button className="btn-small right" waves="light">Register</Button>
							</Col>
						</Row>
					</form>
				</Col>
			</Row>
			)
		}
}

Register.propTypes = {
	registerUser: PropTypes.func.isRequired,
	auth: PropTypes.object.isRequired,
	errors: PropTypes.object.isRequired,
}

const mapStateToProps = (state) => ({
	auth: state.auth,
	errors: state.errors
});

export default connect(mapStateToProps, {registerUser })(withRouter(Register));