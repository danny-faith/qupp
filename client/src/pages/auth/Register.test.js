import React from 'react';
import { shallow, render, mount } from 'enzyme';
import { create } from "react-test-renderer";
import { Register } from './Register';

describe('Register component', () => {
    let wrapper;
    const mockRegisterUserfn = jest.fn();
    const auth = {
        isAuthenticated: true,
    }
    beforeEach(() => {
        wrapper = shallow(
        <Register
            registerUser={mockRegisterUserfn} 
            auth={auth}
            errors={{}}
            history={[]}
        />)
    });

    
    it('should shallow render correctly', () => {
        const auth = {
            isAuthenticated: false,
        }
        const mockRegisterUserfn = jest.fn();
        const component = shallow(
            <Register
                auth={auth}
                registerUser={mockRegisterUserfn}
                errors={{}}
                history={[]}
            />
        );
        expect(component).toMatchSnapshot();
    });

    describe("Register component (react-test-renderer)", () => {
        test("it shows the expected text when clicked (testing the wrong way!)", () => {
            const component = create(
                <Register 
                    auth={auth}
                    registerUser={mockRegisterUserfn}
                    errors={{}}
                    history={[]}
                />);
            const rootInstance = component.root;
            const register = rootInstance.findByProps({name: "email"});
            console.log(register.props);
            expect(register.props.name).toBe("email");
            
            
            // expect(instance.props.children).toBe("");
        });
    });
    
    describe('When the form is submitted', () => {
        it('should call the mock login function', () => {
        // Ok to use `form` as a selector as there is only one form within the component
         wrapper.find('form').simulate(
            'submit', 
            {preventDefault() {}}
        );
        expect(mockRegisterUserfn.mock.calls.length).toBe(1)
        })
    });

    it('should be called with the email, username, password and password2 in the state as arguments', () => {
        // fill in email field with blah@gmail.com     
        wrapper.find('[name="email"]').simulate(
          'change', 
          {target: 
            {name: 'email', value: 'blah@gmail.com'}
          }
        )
        // fill in username field with blah@gmail.com     
        wrapper.find('[name="username"]').simulate(
          'change', 
          {target: 
            {name: 'username', value: 'JohnSmith'}
          }
        )
        // fill in password field with cats  
        wrapper.find('[name="password"]').simulate(
          'change', 
          {target: 
             {name: 'password', value: 'cats'}
          }
        )
        // fill in password field with cats  
        wrapper.find('[name="password2"]').simulate(
          'change', 
          {target: 
             {name: 'password2', value: 'cats'}
          }
        )
        // simulate form submission   
        // Ok to use `form` as a selector as there is only one form within the component
        wrapper.find('form').simulate(
          'submit', 
          {preventDefault() {}}
        )
        // test to see arguments used after its been submitted 
        expect(mockRegisterUserfn.mock.calls[1][0]).toEqual(
            {
                email: 'blah@gmail.com',
                username: 'JohnSmith',
                password: 'cats',
                password2: 'cats'
            }
        )
    });
});