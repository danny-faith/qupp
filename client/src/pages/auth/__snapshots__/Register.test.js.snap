// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component should shallow render correctly 1`] = `
<Row>
  <Col
    className="offset-s3"
    s={6}
  >
    <h1
      className="center"
    >
      Register
    </h1>
    <form
      noValidate={true}
      onSubmit={[Function]}
    >
      <Row>
        <TextFieldGroup
          label="Email"
          name="email"
          onChange={[Function]}
          type="email"
          value=""
        />
        <TextFieldGroup
          label="Username"
          name="username"
          onChange={[Function]}
          type="text"
          value=""
        />
        <TextFieldGroup
          label="Password"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
        <TextFieldGroup
          label="Confirm password"
          name="password2"
          onChange={[Function]}
          type="password"
          value=""
        />
      </Row>
      <Row>
        <Col
          s={12}
        >
          <Button
            className="btn-small right"
            node="button"
            waves="light"
          >
            Register
          </Button>
        </Col>
      </Row>
    </form>
  </Col>
</Row>
`;
